File: VB/TreeListEditing/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Web.Mvc
Imports TreeListEditing.Models

Namespace TreeListEditing.Controllers
	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult
			Return View(EmployeesDataProvider.GetData())
		End Function

		<ValidateInput(False)> _
		Public Function TreeListPartial() As ActionResult
			Return PartialView("TreeListPartial", EmployeesDataProvider.GetData())
		End Function

		<HttpPost, ValidateInput(False)> _
		Public Function TreeListPartialAddNew(ByVal employee As Employee) As ActionResult
			If ModelState.IsValid Then
				Try
					EmployeesDataProvider.AddEmployee(employee)
				Catch e As Exception
					ViewData("EditNodeError") = e.Message
				End Try
			Else
				ViewData("EditNodeError") = "Please, correct all errors."
			End If
			Return PartialView("TreeListPartial", EmployeesDataProvider.GetData())
		End Function

		<HttpPost, ValidateInput(False)> _
		Public Function TreeListPartialUpdate(ByVal employee As Employee) As ActionResult
			If ModelState.IsValid Then
				Try
					EmployeesDataProvider.UpdateEmployee(employee)
				Catch e As Exception
					ViewData("EditNodeError") = e.Message
				End Try
			Else
				ViewData("EditNodeError") = "Please, correct all errors."
			End If
			Return PartialView("TreeListPartial", EmployeesDataProvider.GetData())
		End Function

		<HttpPost, ValidateInput(False)> _
		Public Function TreeListPartialMove(ByVal employeeID As Integer, ByVal supervisorID As Nullable(Of Integer)) As ActionResult
			EmployeesDataProvider.MoveEmployee(employeeID, supervisorID)
			Return PartialView("TreeListPartial", EmployeesDataProvider.GetData())
		End Function

		<HttpPost, ValidateInput(False)> _
		Public Function TreeListPartialDelete(ByVal employeeID As Integer) As ActionResult
			Try
				EmployeesDataProvider.DeleteEmployee(employeeID)
			Catch e As Exception
				ViewData("EditNodeError") = e.Message
			End Try
			Return PartialView("TreeListPartial", EmployeesDataProvider.GetData())
		End Function
	End Class
End Namespace

File: CS/TreeListEditing/Controllers/HomeController.cs
using System;
using System.Web.Mvc;
using TreeListEditing.Models;

namespace TreeListEditing.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View(EmployeesDataProvider.GetData());
        }

        [ValidateInput(false)]
        public ActionResult TreeListPartial() {
            return PartialView("TreeListPartial", EmployeesDataProvider.GetData());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult TreeListPartialAddNew(Employee employee) {
            if(ModelState.IsValid) {
                try {
                    EmployeesDataProvider.AddEmployee(employee);
                }
                catch(Exception e) {
                    ViewData["EditNodeError"] = e.Message;
                }
            }
            else
                ViewData["EditNodeError"] = "Please, correct all errors.";
            return PartialView("TreeListPartial", EmployeesDataProvider.GetData());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult TreeListPartialUpdate(Employee employee) {
            if(ModelState.IsValid) {
                try {
                    EmployeesDataProvider.UpdateEmployee(employee);
                }
                catch(Exception e) {
                    ViewData["EditNodeError"] = e.Message;
                }
            }
            else
                ViewData["EditNodeError"] = "Please, correct all errors.";
            return PartialView("TreeListPartial", EmployeesDataProvider.GetData());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult TreeListPartialMove(int employeeID, int? supervisorID) {
            EmployeesDataProvider.MoveEmployee(employeeID, supervisorID);
            return PartialView("TreeListPartial", EmployeesDataProvider.GetData());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult TreeListPartialDelete(int employeeID) {
            try {
                EmployeesDataProvider.DeleteEmployee(employeeID);
            }
            catch(Exception e) {
                ViewData["EditNodeError"] = e.Message;
            }
            return PartialView("TreeListPartial", EmployeesDataProvider.GetData());
        }
    }
}